core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/ListVariableListener.java

```diff
@@ -1,19 +1,94 @@
 package org.optaplanner.core.impl.domain.variable;
 
+import org.optaplanner.core.api.domain.entity.PlanningEntity;
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
 import org.optaplanner.core.api.domain.variable.AbstractVariableListener;
+import org.optaplanner.core.api.domain.variable.PlanningListVariable;
 import org.optaplanner.core.api.score.director.ScoreDirector;
 
+/**
+ * A listener sourced on a {@link PlanningListVariable}.
+ * <p>
+ * Changes shadow variables when a genuine source list variable changes.
+ * <p>
+ * Important: it must only change the shadow variable(s) for which it's configured!
+ * It should never change a genuine variable or a problem fact.
+ * It can change its shadow variable(s) on multiple entity instances
+ * (for example: an arrivalTime change affects all trailing entities too).
+ * <p>
+ * It is recommended that implementations be kept stateless.
+ * If state must be implemented, implementations may need to override the default methods
+ * ({@link #resetWorkingSolution(ScoreDirector)}, {@link #close()}).
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ * @param <Entity_> @{@link PlanningEntity} on which the source variable is declared
+ */
 public interface ListVariableListener<Solution_, Entity_> extends AbstractVariableListener<Solution_, Entity_> {
 
-    void beforeListVariableElementAdded(ScoreDirector<Solution_> scoreDirector, Entity_ entity, int index);
-
-    void afterListVariableElementAdded(ScoreDirector<Solution_> scoreDirector, Entity_ entity, int index);
-
-    void beforeListVariableElementRemoved(ScoreDirector<Solution_> scoreDirector, Entity_ entity, int index);
-
-    void afterListVariableElementRemoved(ScoreDirector<Solution_> scoreDirector, Entity_ entity, int index);
+    /**
+     * The listener must unset all shadow variables it is responsible for when an element is unassigned from the source list
+     * variable. For example, a {@code Task}'s {@code startTime} shadow variable must be reset to {@code null} after a task
+     * is unassigned from {@code Employee.tasks} when the move that assigned it there is undone during Construction Heuristic
+     * phase.
+     *
+     * @param scoreDirector score director
+     * @param element the unassigned element
+     */
+    void afterListVariableElementUnassigned(ScoreDirector<Solution_> scoreDirector, Object element);
 
+    /**
+     * Tells the listener that some elements within the range starting at {@code fromIndex} (inclusive) and ending at
+     * {@code toIndex} (exclusive) will change.
```

Comment:
The most surprising part of this code is that the before and after calls might have different ranges.
Let's make sure the javadoc clearly mentions that. I would make it the second paragraph.